### PR TITLE
[Merged by Bors] - feat(RepresentationTheory/GroupCohomology): add some API for group cohomology in low degree

### PR DESCRIPTION
**Pull request by 101damnations**
---

In `RepresentationTheory.GroupCohomology.Basic`, we define the `n`th group cohomology of a `k`-linear `G`-representation `A` to be the cohomology of a complex `inhomogeneousCochains A`, whose objects are `(Fin n → G) → A`; this is
unnecessarily unwieldy in low degree. So this PR adds a file `LowDegree` giving better definitions for group cohomology for `n = 0, 1, 2`, denoted `H0, H1, H2` (as opposed to the existing definition `groupCohomology A n`.)

This would mean there are two definitions of group cohomology for `n = 0, 1, 2`, which is probably bad. It also means the documentation in `LowDegree` is quite unclear. **I think maybe I should change `groupCohomology` to be this PR's definitions for `n = 0, 1, 2` and its original definition for higher `n`, but that's still a bit weird... What do people think?**

This PR probably shouldn't be any longer than it is but the next addition will be an identification between `groupCohomology A n` and `Hn A` for `n = 0, 1, 2` as they're defined now. This should be added whether or not I kind of collapse the 2 definitions into  1 as described above. In Lean 3 this identification was a slow isomorphism. I will refactor it for the next PR, guided by the homology refactor.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by joelriou** [comment]
I think we should not change the definition of `groupCohomology`. It will be more than sufficient to provide the suitable additive equivalences in a future PR.

**Comment by digama0** [comment]
Note: I have pushed an update to the lean toolchain because this PR was on a [buggy version](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Potential.20data.20loss.20from.20.60lake.20clean.60.20with.204.2E2.2E0-rc2.2F3/near/397875701) of the toolchain. **WARNING**: checking out old commits of this PR using v4.2.0-rc2 or v4.2.0-rc3 can cause `lake clean` to delete your mathlib folder! If you need to do so, make sure to delete `lakefile.olean` manually before running any `lake` commands.

**Comment by joelriou** [comment]
I am trying to think about how to proceed next so that it is the most convenient wrt to the homology refactor. I have already PRed a significant part of the material about the homology theory of `ShortComplex`. Then, I may start to consider doing what I suggested in:
https://leanprover.zulipchat.com/#narrow/stream/335062-homology/topic/Homology.20refactor/near/384208971
which is, do a PR which basically just renames the current homology theory of homological complex as `homology'`, etc, instead of `homology`. Then, I would introduce the new homology API for homological complexes (i.e. redefine `HomologicalComplex.homology`), so that you may benefit from the new homology API in order to get the comparison isos in low degrees. (For example, for the comparison of group cohomology in degree `1`, it suffices to show that the short complex consisting of the standard cochains in degrees `0`, `1` and `2` identifies to the short complex given by the more explicit formulas in low degrees, and this shall mostly be given by `ShortComplex.isoMk`, `oneCochainsLinearEquiv`, etc, and the compatibilities with the differentials you have obtained here.) As this may still take a little time (a few weeks hopefully?), I may suggest you postpone PRing the isomorphisms between the two "definitions" in low degrees, but computations which involve low degree cohomology (e.g. cyclic groups, Hilbert 90, etc) would be very much welcome!

I think this PR shall mostly look good to me as soon as you have merged with master (in order to take into account #6029).

**Comment by 101damnations** [comment]
Thank you very much! That makes sense; I'm happy to wait to add the isomorphisms, and PR other things in the meantime.

File: Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+  have := ((mem_twoCocycles_iff f.1).1 f.2 (1, (1, g))).symm
+  simpa only [map_one, LinearMap.one_apply, one_mul, add_right_inj, this]
+
+-- should maybe be the other way around, with a different name?
+theorem twoCocycles_map_one_snd (f : twoCocycles A) (g : G) :
+    f.1 (g, 1) = A.ρ g (f.1 (1, 1)) := by
+  have := (mem_twoCocycles_iff f.1).1 f.2 (g, (1, 1))
+  simpa only [mul_one, add_left_inj, this]
+
+-- no idea for naming
```
Comment (joelriou) [inline Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean:254]: I think the statements and names are ok. You may remove the comments on lies 248/254.

bors d+

**Pull request closed without merging**